src/llfs/simulated_log_device_storage.cpp

```diff
@@ -0,0 +1,539 @@
+    StatusOr<std::function<void()>> handler = this->queue_.await_next();
+    if (!handler.ok()) {
+      if (handler.status() == batt::StatusCode::kPoke) {
+        continue;
+      }
+
+      if (handler.status() == batt::StatusCode::kClosed) {
+        break;
+      }
+
+      return handler.status();
+    }
+
+    try {
+      (*handler)();
+    } catch (...) {
+      LLFS_LOG_WARNING() << "Simulation handler threw exception!";
+    }
+  }
+  LLFS_VLOG(1) << "(id=" << this->id_ << ") EphemeralState::run_event_loop() exiting";
+
+  return batt::OkStatus();
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+void SimulatedLogDeviceStorage::EphemeralState::reset_event_loop()
+{
+  LLFS_VLOG(1) << "(id=" << this->id_ << ") EphemeralState::reset_event_loop()"
+               << BATT_INSPECT(this->stopped_);
+  this->stopped_.store(false);
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+void SimulatedLogDeviceStorage::EphemeralState::stop_event_loop()
+{
+  LLFS_VLOG(1) << "(id=" << this->id_ << ") EphemeralState::reset_event_loop()"
+               << BATT_INSPECT(this->stopped_);
+  this->stopped_.store(true);
+  this->queue_.poke();
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+void SimulatedLogDeviceStorage::EphemeralState::post_to_event_loop(
+    std::function<void(StatusOr<i32>)>&& handler)
+{
+  LLFS_VLOG(1) << "(id=" << this->id_ << ") EphemeralState::post_to_event_loop(handler) entered"
+               << BATT_INSPECT(this->simulation_.is_running());
+
+  this->on_work_started(POST_TO_EVENT_LOOP);
+
+  auto runnable = [shared_this = batt::shared_ptr_from(this),
+                   handler = BATT_FORWARD(handler)]() mutable {
+    auto* const this_ = shared_this.get();
+
+    LLFS_VLOG(1) << "(id=" << this_->id_
+                 << ") EphemeralState::post_to_event_loop(handler) invoked runnable";
+
+    // We don't need to capture `shared_this` in the completion handler we push to the queue, since
+    // either the work count will keep the event loop task running (and if the log driver doesn't
+    // join the even loop task, this is a serious error we don't want to mask) *or* the completion
+    // handler will never run, in which case a dangling pointer doesn't matter (if a tree falls in a
+    // forest...)
+    //
+    this_->queue_
+        .push([this_, handler]() mutable {
+          auto on_scope_exit = batt::finally([&] {
+            this_->on_work_finished(POST_TO_EVENT_LOOP);
+          });
+          BATT_FORWARD(handler)(StatusOr<i32>{0});
+        })
+        .IgnoreError();
+  };
+
+  if (this->simulation_.is_running()) {
+    this->simulation_.post(std::move(runnable));
+  } else {
+    runnable();
+  }
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+Status SimulatedLogDeviceStorage::EphemeralState::read_all(i64 offset, MutableBuffer buffer)
```

Comment:
Sure... made the change.  It's probably not too bad for MutableBuffer and ConstBuffer since these are just simple structs that wrap a void* and size_t.